### PR TITLE
Check `unreserve` and `transfer` returnvalues in debug code

### LINKED ISSUE #8106: Replace `let _ = ...` in the Runtime with `debug_assert!`
Right now, many pallets in FRAME use `let _ = ...` for situations where we assume that a function cannot fail or where we can do nothing about a failure.

Some examples are:
* Trying to unreserve too much balance
* Over slashing
* Check then execute pattern on functions that return a result
* etc...

While it is possible that such states will appear in the actual runtime environment (for example a user is slashed, eating into reserved balance, and then unreserve would actually be more than available), within tests, this shouldn't really happen.

As such we should replace the following patterns in our codebase:

```
// It really should be reserved; there's not much we can do here on fail.
let _ = CurrencyOf::<T>::unreserve(&who, amount);
```

in favor of:

```
// It really should be reserved; there's not much we can do here on fail.
let err_amount = CurrencyOf::<T>::unreserve(&who, amount);
debug_assert!(err_amount.is_zero());
```

Or:

```
// Not much we can do if this fails...
let _ = T::Currency::transfer(&Self::account_id(), &winner, lottery_balance, KeepAlive);
```

for

```
// Not much we can do if this fails...
let res = T::Currency::transfer(&Self::account_id(), &winner, lottery_balance, KeepAlive);
debug_assert!(res.is_ok());
```

### PR DESCRIPTION
**Pull request by hirschenberger**
fixes some of #8106

Thank you for your Pull Request!

Before you submitting, please check that:

- [ ] You added a brief description of the PR, e.g.:
  - What does it do?
  - What important points reviewers should know?
  - Is there something left for follow-up PRs?
- [ ] You labeled the PR appropriately if you have permissions to do so:
  - [ ] `A*` for PR status (**one required**)
  - [ ] `B*` for changelog (**one required**)
  - [ ] `C*` for release notes (**exactly one required**)
  - [ ] `D*` for various implications/requirements
  - [ ] Github's project assignment
- [ ] You mentioned a related issue if this PR related to it, e.g. `Fixes #228` or `Related #1337`.
- [ ] You asked any particular reviewers to review. If you aren't sure, start with GH suggestions.
- [ ] Your PR adheres to [the style guide](https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md)
  - In particular, mind the maximal line length of 100 (120 in exceptional circumstances).
  - There is no commented code checked in unless necessary.
  - Any panickers have a proof or removed.
- [ ] You bumped the runtime version if there are breaking changes in the **runtime**.
- [ ] You updated any rustdocs which may have changed
- [ ] Has the PR altered the external API or interfaces used by Polkadot? Do you have the corresponding Polkadot PR ready?

Refer to [the contributing guide](https://github.com/paritytech/substrate/blob/master/docs/CONTRIBUTING.adoc) for details.

After you've read this notice feel free to remove it.
Thank you!

✄ -----------------------------------------------------------------------------

### CODE REVIEW

**Comment by hirschenberger** [comment]
Breaks API? Strange...

File: frame/bounties/src/lib.rs [hunk 1]
```diff
@@ -435,7 +435,9 @@ decl_module! {
 								} else {
 									// Else this is the curator, willingly giving up their role.
 									// Give back their deposit.
-									let _ = T::Currency::unreserve(&curator, bounty.curator_deposit);
+									let err_amount = T::Currency::unreserve(&curator,
+										bounty.curator_deposit);
```
Comment (kianenigma) [inline frame/bounties/src/lib.rs:439]:
```suggestion
									let err_amount = T::Currency::unreserve(&curator, bounty.curator_deposit);
```

Comment (kianenigma) [reply frame/bounties/src/lib.rs:439]: does it exceed 100? the both `curator` and `boundy` should break

File: frame/society/src/lib.rs [hunk 2]
```diff
@@ -1402,7 +1404,9 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
 					Self::bump_payout(winner, maturity, total_slash);
 				} else {
 					// Move the slashed amount back from payouts account to local treasury.
-					let _ = T::Currency::transfer(&Self::payouts(), &Self::account_id(), total_slash, AllowDeath);
+					let res = T::Currency::transfer(&Self::payouts(), &Self::account_id(),
+						total_slash, AllowDeath);
```
Comment (kianenigma) [inline frame/society/src/lib.rs:1408]:
```suggestion
					let res = T::Currency::transfer(&Self::payouts(), &Self::account_id(),
						total_slash, AllowDeath);
```

bad format.

File: frame/society/src/lib.rs [hunk 3]
```diff
@@ -1413,7 +1417,9 @@ impl<T: Config<I>, I: Instance> Module<T, I> {
 
 				// this should never fail since we ensure we can afford the payouts in a previous
 				// block, but there's not much we can do to recover if it fails anyway.
-				let _ = T::Currency::transfer(&Self::account_id(), &Self::payouts(), total_payouts, AllowDeath);
+				let res = T::Currency::transfer(&Self::account_id(), &Self::payouts(),
+					total_payouts, AllowDeath);
```
Comment (kianenigma) [inline frame/society/src/lib.rs:1421]: likewise

**Comment by hirschenberger** [comment]
Fixed formatting, I was not sure if these lines longer than 100 chars count as *exceptional circumstances* from the coding rules.

File: frame/bounties/src/lib.rs (hunk 1 above)
Comment (hirschenberger) [reply frame/bounties/src/lib.rs]: Oh, just changed it to non-breaking bc. of your former comment. Should I change it back?

File: frame/bounties/src/lib.rs [hunk 4]
```diff
@@ -548,9 +549,13 @@ decl_module! {
 					let balance = T::Currency::free_balance(&bounty_account);
 					let fee = bounty.fee.min(balance); // just to be safe
 					let payout = balance.saturating_sub(fee);
-					let _ = T::Currency::unreserve(&curator, bounty.curator_deposit);
-					let _ = T::Currency::transfer(&bounty_account, &curator, fee, AllowDeath); // should not fail
-					let _ = T::Currency::transfer(&bounty_account, &beneficiary, payout, AllowDeath); // should not fail
+					let err_amount = T::Currency::unreserve(&curator, bounty.curator_deposit);
+					debug_assert!(err_amount.is_zero());
+					let res = T::Currency::transfer(&bounty_account, &curator, fee, AllowDeath); // should not fail
```
Comment (hirschenberger) [inline frame/bounties/src/lib.rs:554]: Also these Lines exceed 100, but they did before bc. of the comment, so I didn't change it.

**Comment by hirschenberger** [comment]
@kianenigma How should we continue here? Should I change it back to cap the linelengths to 100?

**Comment by kianenigma** [comment]
> @kianenigma How should we continue here? Should I change it back to cap the linelengths to 100?

as long as you pass the line-width CI and there is no exceptionally ugly line in the manual review, you are good to go!

**Review by thiolliere — COMMENTED** [review]
changes look good, but it seems it catch an error in a test, needs to be solved:

```
test tests_composite::emit_events_with_reserve_and_unreserve ... FAILED
test tests_local::emit_events_with_reserve_and_unreserve ... FAILED
```

**Comment by hirschenberger** [comment]
Ok, should be fixed now, also no use for debug_asserts in tests. Changed to usual asserts.

File: frame/balances/src/tests.rs [hunk 5]
```diff
@@ -692,15 +692,15 @@ macro_rules! decl_tests {
 					);
 
 					System::set_block_number(3);
-					let _ = Balances::unreserve(&1, 5);
+					assert!(Balances::unreserve(&1, 5).is_zero());
```
Comment (shawntabrizi) [inline frame/balances/src/tests.rs:695]: I see some `let _ = ...` above.

Looks like they should be wrapped in `assert_ok!(..)`

Comment (hirschenberger) [reply frame/balances/src/tests.rs:695]: I can do it, it is not directly related to this PR but I can change it as I change this file.

Comment (hirschenberger) [reply frame/balances/src/tests.rs:695]: So I'm wrapping all Result-returning functions, but I can leave out the ones that are unfailable ones like `slash` or `deposit_creating` right?

**Comment by thiolliere** [comment]
> Ok, should be fixed now, also no use for debug_asserts in tests. Changed to usual asserts.

ah in general I prefer not to have `force-pushed` so that I can see how the issue was address. And be able to review only the new changes.
How did you changed it ?

**Review by thiolliere — APPROVED** [review]

**Comment by hirschenberger** [comment]
> > Ok, should be fixed now, also no use for debug_asserts in tests. Changed to usual asserts.
> 
> ah in general I prefer not to have `force-pushed` so that I can see how the issue was address. And be able to review only the new changes.
> How did you changed it ?

Sorry, I thought force-pushing is the common way here to keep the commit-count low and consistent.

The failing unittest was [here](https://github.com/paritytech/substrate/pull/8398/commits/2ae26f0057a165a1ca918d6d238678098732bc6b#diff-71390538a8f10951b668bc36f72a1c030e4b68630fad36f72578190ba65e0cbbR703).

It was an `unreserve` which was expected to **not** return Zero.

**Comment by thiolliere** [comment]
great, all good then.

>  I thought force-pushing is the common way here to keep the commit-count low and consistent.

Yes there is pro and cons for both strategies

File: frame/balances/src/tests.rs (hunk 5 above)
Comment (shawntabrizi) [reply frame/balances/src/tests.rs:695]: yup!

Comment (shawntabrizi) [reply frame/balances/src/tests.rs:695]: ill fix it for you real quick, just a nit pick

**Review by shawntabrizi — APPROVED** [review]
Ah, I see now you are only hunting specific functions.

This is great for now. Will need to hunt for other places where this may happen in a future PR.

**Comment by shawntabrizi** [comment]
bot merge

**Pull request merged**